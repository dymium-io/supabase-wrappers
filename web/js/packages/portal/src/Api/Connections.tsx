import Alert from 'react-bootstrap/Alert'
import * as types from '@dymium/common/Types/Common'
import * as internal from '@dymium/common/Types/Internal'
import * as com from '../Common'

export function getConnections(setSpinner, setConns, setAlert, remap:internal.ConnectionMap|undefined, onSuccess) {
    setSpinner(true)
    setConns([])
    com.sendToServer("GET", "/api/getconnections",
        null, "",
        resp => {

            resp.json().then(_js => {
                let js = types.ConnectionResponse.fromJson(_js)
                if (js.status !== "OK") {
                    setAlert(
                        <Alert variant="danger" onClose={() => setAlert(<></>)} dismissible>
                            Error retrieving connections: {js.errormessage} { }
                        </Alert>
                    )
                    setTimeout(() => setSpinner(false), 500)
                    return
                }
            
                let cc = js.data.map(x => {
                    let ob = types.ConnectionRecord.fromJson({
                        id: x.id,
                        credid: x.credid,
                        dbtype: x.dbtype,
                        name: x.name,
                        dbname: x.dbname,
                        address: x.address,
                        port: x.port,
                        description: x.description,
                        useTLS: x.useTLS,

                    })
                    if(undefined != remap && x.name != null)
                        remap[x.name] = ob
                    return ob
                })

                setConns(cc)
                if(onSuccess != undefined) {
                    onSuccess()
                }
            })

            setTimeout(() => setSpinner(false), 500)
        },
        resp => {
            console.log("on error")
            setSpinner(false)
        },
        error => {
            console.log("on exception: " + error)
            setSpinner(false)
        })
}