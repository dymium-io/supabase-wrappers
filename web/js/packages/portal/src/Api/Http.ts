
export function sendToServer(method: string, url: string,
    headers, body: string, onsuccess, onfailure, onexception) {
    let token = window.sessionStorage.getItem("Session");

    if (token === null) {
        onfailure(null)
        return;
    }
    let _headers = {
        Authorization: "Bearer " + token,
        Cache: "no-cache",
    }
    if (headers != null) {
        _headers = {
            ..._headers,
            ...headers
        }
    }
    let params: any = {
        method,
        headers: _headers
    }
    if (body != "") {
        params = { ...params, body: body }
    }
  

    fetch(url, params
    ).then(response => {

        if (!response.ok) {
            onfailure(response)
        } else {
            onsuccess(response)
        }
    }).catch(onexception)
}
