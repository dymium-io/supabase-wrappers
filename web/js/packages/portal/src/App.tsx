import React, { useEffect } from 'react';
import LandingPage from './LandingPage'
import AuthenticatedApp from './App/AuthenticatedApp'
import Error404 from "./Error404"
import Logout from "./Logout"
import Error from "./Error"
import Invitation from "./Invitation"
import Sticky from "./App/Sticky"
import Dashboard from "./App/Dashboard"
import Customers from "./App/Customers"
import Connections from "./App/Connections"
import Datascopes from "./App/Datascopes"
import Groups from "./App/Groups"
import Rules from "./App/Rules"
import TestSQL from "./App/TestSQL"
import * as com from './Common'

import "@fontsource/roboto";
import 'bootstrap/dist/css/bootstrap.min.css';
import 'react-bootstrap-typeahead/css/Typeahead.css';

import { BrowserRouter, Routes, Route } from "react-router-dom";
//import Helmet from "react-helmet";

import '@dymium/common/App.scss';


function App() {
  return (
    <div className="App" id="app">

      <BrowserRouter>
        <Routes>

          <Route  path="/" element={<LandingPage />} />

          <Route path="/app/" element={<AuthenticatedApp />} >
            <Route path="/app/dashboard" element={<Dashboard />} />
            <Route path="/app/connections" element={<Connections />} />     
            <Route path="/app/datascopes" element={<Datascopes />} />                
            <Route path="/app/groups" element={<Groups />} />    
            <Route path="/app/rules" element={<Rules />} />         
            <Route path="/app/test" element={<TestSQL />} />                                             
            <Route path="/app/" element={<Sticky />} />          
          </Route>
          <Route  path="/app/logout" element={<Logout/>} />
          <Route path="/app/error" element={<Error />} />
          <Route path="/login" element={<Invitation />} />
          <Route path="*" element={<Error404 />} />
        </Routes>
      </BrowserRouter>

    </div>
  );
}

export default App;
