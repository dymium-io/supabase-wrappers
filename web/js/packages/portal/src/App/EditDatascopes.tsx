import React, { useEffect, useState, useRef } from 'react';
import Form from 'react-bootstrap/Form'
import Button from 'react-bootstrap/Button'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Offcanvas from '@dymium/common/Components/Offcanvas'
import Alert from 'react-bootstrap/Alert'
import Spinner from '@dymium/common/Components/Spinner'
import Modal from 'react-bootstrap/Modal'
import BootstrapTable from 'react-bootstrap-table-next';
import paginationFactory from 'react-bootstrap-table2-paginator';
import ToolkitProvider, { Search } from 'react-bootstrap-table2-toolkit/dist/react-bootstrap-table2-toolkit';
import AddTable from './AddTable'
import { useAppDispatch, useAppSelector } from './hooks'
import { setSelectedDatascopeDefault } from '../Slices/menuSlice'
import * as com from '../Common'
import DatascopeForm from './DatascopeForm'
import * as internal from '@dymium/common/Types/Internal'
import * as types from '@dymium/common/Types/Common'
import * as capi from '../Api/Connections'
import * as http from '@dymium/common/Api/Http'

const { SearchBar, ClearSearchButton } = Search;

let remap = new internal.ConnectionMap();

export default function EditDatascopes() {
    const [spinner, setSpinner] = useState(false)
    let [conns, setConns] = useState<internal.Connection[]>([])
    const [alert, setAlert] = useState<JSX.Element>(<></>)
    const [showdelete, setShowdelete] = useState(false)
    const [slatedToDelete, setSlatedToDelete] = useState("")
    const [datascopes, setDatascopes] = useState<types.DatascopeIdName[]>([])
    const [selectedDatascope, setSelectedDatascope] = useState("")
    const [selectedDatascopeDetails, setSelectedDatascopeDetails] = useState<types.Datascope | null>(null)

    const [initialTables, setInitialTables] = useState<internal.TablesMap>({})
    const [validated, setValidated] = useState(false)
    let form = useRef<HTMLFormElement>(null)

    let setSDRef = useRef(setSelectedDatascope)

    const [showOffcanvas, setShowOffcanvas] = useState<boolean>(false)
    const [table, setTable] = useState<internal.TableScope>({ schema: "", table: "" })
    const [dbname, setDbname] = useState("")
    const [datascope, setDatascope] = useState({})
    const [currentConnectionId, setCurrentConnectionId] = useState("")
    const [currentConnectionType, setCurrentConnectionType] = useState("")
    const [showOffhelp, setShowOffhelp] = useState(com.isInstaller())

    let t = useAppSelector((state) => {

        return state.reducer.selectedDatascope
    }
    )
    const appDispatch = useAppDispatch()
    let nameById = id => {
        for(let i = 0; i < datascopes.length; i++) {
            if(id === datascopes[i].id)
                return datascopes[i].name
        }
        return ""
    }
    let onEdit = (id) => {
        return e => {
            console.log(id)
            setSDRef.current(id)
      
            appDispatch(setSelectedDatascopeDefault(id))
        }
    }
    let onDelete = (id) => {
        return e => {
            setSlatedToDelete(id)
            setShowdelete(true)
        }
    }
    let columns = [
        {
            dataField: 'id',
            text: 'id',
            hidden: true,
        },

        {
            dataField: 'name',
            text: 'Name:',
            sort: true,
        },
        {
            text: 'Edit',
            dataField: 'edit',
            isDummyField: false,
            formatter: (cell, row, rowIndex, formatExtraData) => {
                return <i className="fas fa-edit ablue" onClick={onEdit(row["id"])} role="button"></i>
            },
            //formatExtraData: { hoverIdx: this.state.hoverIdx },
            headerStyle: { width: '50px' },
            style: { height: '30px' },
            align: 'center'
        },
        {
            text: 'Delete',
            dataField: 'delete',
            isDummyField: true,
            formatter: (cell, row, rowIndex, formatExtraData) => {
                return <i className="fas fa-trash ablue" onClick={onDelete(row["id"])} role="button"></i>
            },
            //formatExtraData: { hoverIdx: this.state.hoverIdx },
            headerStyle: { width: '90px' },
            style: { height: '30px' },
            align: 'center'
        }
    ]

    let selectRow = {
        mode: 'radio',
        //clickToSelect: true,
        style: { backgroundColor: 'rgba(0, 151, 206, 0.3)' },
        selected: [t],
        onSelect: (row, isSelect, rowIndex, e) => {
            console.log("in onselect", row["id"])
            setSDRef.current(row["id"])
            appDispatch(setSelectedDatascopeDefault(row["id"]))
        },
    };
    let reload = () => { 
        capi.getConnections(setSpinner, setConns, setAlert, remap, () => {

            com.getDatascopes(setSpinner, setAlert, setDatascopes, () => {
                setSelectedDatascope(t)
            })
        })
    }
    useEffect(() => {
        reload()
    }, [])
    useEffect(() => {
        if (selectedDatascope === "" )
            return
        let body = types.DatascopeId.fromJson({ id: selectedDatascope }).toJson()
        setSpinner(true)
        http.sendToServer("POST", "/api/getdatascopedetails",
            null, body,
            resp => {
                resp.json().then(fjs => {
                    let ojs = types.DatascopeInfoStatus.fromJson(fjs)
                    if (ojs.status !== "OK") {
                        setSpinner(false)
                        if(ojs.status !== "sql: no rows in result set")
                            setAlert(
                                <Alert variant="success" onClose={() => setAlert(<></>)} dismissible>
                                    {ojs.errormessage}
                                </Alert>
                            )
                        setSelectedDatascope("")
                        appDispatch(setSelectedDatascopeDefault(""))
                        return
                    }
                    let js = ojs.record
                    if (js == null) {
                        setSpinner(false)
                        setAlert(
                            <Alert variant="success" onClose={() => setAlert(<></>)} dismissible>
                                No record returned
                            </Alert>
                        )
                    }

                    setSelectedDatascopeDetails(js)
                    if (js != null)
                        setDbname(js.name)
                    let ob = {}
                    if (js != null && js.records != null) {
                        js.records.forEach(r => {
                            if (ob[r.connection] === undefined) {
                                ob[r.connection] = {}
                            }
                            let key = r.schema + "." + r.table
                            if (ob[r.connection][key] === undefined) {
                                ob[r.connection][key] = {}
                                ob[r.connection][key]["connection"] = r.connection
                                ob[r.connection][key]["schema"] = r.schema
                                ob[r.connection][key]["table"] = r.table
                                ob[r.connection][key]["tablescope"] = []
                            }
                            let line = {}
                            line["id"] = r.id
                            line["connection"] = r.connection
                            line["action"] = r.action
                            line["name"] = r.col
                            line["position"] = r.position
                            line["reference"] = r.reference
                            line["semantics"] = r.semantics
                            line["dflt"] = r.dflt
                            line["isnullable"] = r.isnullable
                            line["possibleActions"] = r.possibleActions
                            line["typ"] = r.typ
                            ob[r.connection][key]["tablescope"].push(line)

                        })
                    }
                    setInitialTables(ob)

                })
                setSpinner(false)
                capi.getConnections(setSpinner, setConns, setAlert, remap, () => {
                    com.getDatascopes(setSpinner, setAlert, setDatascopes, () => {
                        setSelectedDatascope(t)
                    })
                })
            },
            resp => {
                console.log("on error")
                setSpinner(false)
                resp != null && resp.text().then(t =>
                    setAlert(
                        <Alert variant="danger" onClose={() => setAlert(<></>)} dismissible>
                            {t}
                        </Alert>
                    )                    
                )
            },
            error => {
                console.log("on exception: " + error)
                setSpinner(false)
            })

    }, [selectedDatascope]
    )

    let updateConnection = () => {
        let retarray: internal.DatascopeRecord[] = []
        Object.keys(datascope).forEach(connection => {
            let conn = datascope[connection]
            Object.keys(conn).forEach(schematable => {
                let st = conn[schematable]
                // connection, schema, table, tablescope[typ, semantics, name, position, reference, action]

                st.tablescope.forEach(ts => {
                    let ob: internal.DatascopeRecord = {
                        connection: st.connection, schema: st.schema, table: st.table,
                        typ: ts.typ, position: ts.position, reference: ts.reference, action: ts.action,
                        col: ts.name, semantics: ts.semantics, dflt: ts.dflt, isnullable: ts.isnullable, 
                        possibleActions: ts.possibleActions
                    }
                    retarray.push(ob)
                })
            })

        })
        // now do send
        setSpinner(true)
        let retob: internal.DataScope = { name: dbname, records: retarray }
        let body = JSON.stringify(retob)
        http.sendToServer("POST", "/api/updatedatascope",
            null, body,
            resp => {

                resp.json().then(js => {
                    if (js.status === "OK") {
                        setAlert(
                            <Alert variant="success" onClose={() => setAlert(<></>)} dismissible>
                                Ghost Database {dbname} updated successfully!
                            </Alert>
                        )
                        capi.getConnections(setSpinner, setConns, setAlert, remap, () => {

                            com.getDatascopes(setSpinner, setAlert, setDatascopes, () => {
                                setSelectedDatascope(t)
                            })
                        })
                    } else {
                        setAlert(
                            <Alert variant="danger" onClose={() => setAlert(<></>)} dismissible>
                                Error updating {dbname}:  {js.errormessage}!
                            </Alert>
                        )
                    }
                })
                setSpinner(false)
            },
            resp => {
                console.log("on error")
                setSpinner(false)
                resp != null && resp.text().then(t =>
                    setAlert(
                        <Alert variant="danger" onClose={() => setAlert(<></>)} dismissible>
                            Error updating {dbname}:  {t}
                        </Alert>
                    ))
            },
            error => {
                console.log("on exception: " + error)
                setSpinner(false)
            })

    }
    let addTableR: any = useRef(null)
    let onAddTableRef = (theref) => {
        addTableR.current = theref
    }
    let onAddTable = (table) => {
        setShowOffcanvas(false)
        if (addTableR.current !== undefined && addTableR.current.current) {

            addTableR.current.current(table)
        }
    }
    let handleSubmit = event => {
        if (form.current == null) {
            return false
        }
        if (form.current.reportValidity() === false) {
            event.preventDefault();
            setValidated(true)
            return false
        }

        event.preventDefault();
        setValidated(false)
        event.stopPropagation();

        updateConnection()

        return false
    }
    let onTablesMapUpdate = (t: internal.TablesMap) => {
        setDatascope(t)
    }

    let onEditTable = (t: internal.TableScope) => {
        setTable(t)
        setShowOffcanvas(true)
    }

    let addNewTable = (id: string, dbtype:string, schema?: string, table?: string) => {
        setCurrentConnectionId(id)
        setCurrentConnectionType(dbtype)
        if (schema === undefined || table === undefined)
            setTable({ schema: "", table: "" })
        else
            setTable({ schema, table })
        setShowOffcanvas(true)
    }
    let deleteDatascope = () => {
        setShowdelete(false)
        setSpinner(true)
        let retob: types.DatascopeIdName = new types.DatascopeIdName()
        retob.id = slatedToDelete
        retob.name = nameById(slatedToDelete)
        let body = retob.toJson()

        http.sendToServer("POST", "/api/deletedatascope",
            null, body,
            resp => {

                resp.json().then(js => {
                    if (js.status === "OK") {
                        setAlert(
                            <Alert variant="success" onClose={() => setAlert(<></>)} dismissible>
                                Ghost Database {dbname} updated successfully!
                            </Alert>
                        )
                        capi.getConnections(setSpinner, setConns, setAlert, remap, () => {

                            com.getDatascopes(setSpinner, setAlert, setDatascopes, () => {
                                setSelectedDatascope(t)
                            })
                        })
                    } else {
                        setAlert(
                            <Alert variant="danger" onClose={() => setAlert(<></>)} dismissible>
                                Error updating {dbname}:  {js.errormessage}
                            </Alert>
                        )
                    }
                })
                setSpinner(false)
            },
            resp => {
                console.log("on error")
                setSpinner(false)
                resp != null && resp.text().then(t =>
                    setAlert(
                        <Alert variant="danger" onClose={() => setAlert(<></>)} dismissible>
                            Error updating {dbname}:  {t}
                        </Alert>
                    )
                )
            },
            error => {
                console.log("on exception: " + error)
                setSpinner(false)
            })

    }
    let onDeleteConnection = (c:string) => {
        delete datascope[c]
        setDatascope(datascope)
    }
    return (
        <div className=" text-left">
            {alert}            
            <Offcanvas show={showOffcanvas} onClose={(e) => { setShowOffcanvas(false) }}
                title={table["connection"] === undefined ? "Register table" : "Edit table"}>
                {showOffcanvas &&
                    <AddTable onHide={() => { setShowOffcanvas(false) }} onAlert={setAlert} onAddTable={onAddTable} table={table} currentConnectionType={currentConnectionType}  connectionId={currentConnectionId} />
                }
            </Offcanvas>
            <Offcanvas modal={false} width={300} show={showOffhelp} onClose={(e) => { setShowOffhelp(false) }}>
                <h5>Editing Ghost Database</h5>
                <div className="mb-3">
                    This page allows to edit or remove a Ghost Database.
                </div>
                <div className="mb-3">
                    The same interface as for adding a Ghost Database is used. You select Data Sources and tables that you want to expose
                </div>
            </Offcanvas>            
            <Modal centered show={showdelete} onHide={() => setShowdelete(false)} data-testid="modal-delete">
                <Modal.Header closeButton>
                    <Modal.Title>Delete Datascope {nameById(slatedToDelete)}?</Modal.Title>
                </Modal.Header>
                <Modal.Body>Are you sure you want to remove the Datascope? This operation is irreversible.</Modal.Body>
                <Modal.Footer>
                    <Button variant="danger"  role="button" id="Delete" data-testid="Delete"
                         aria-label={"Delete"}
                        onClick={() => {
                            deleteDatascope()
                    }
                    }>Delete</Button> <Button variant="dymium" onClick={() => {
                        setShowdelete(false)
                    }}>Cancel</Button>
                </Modal.Footer>
            </Modal>
            <div className=" text-left">
                {/*
                <Row> <Col xs="auto">
                    <Form.Group className="mb-3" controlId="connection" >
                        <Form.Label >Available Ghost Databases</Form.Label>
                        <Form.Control as="select" size="sm"
                            onChange={e => {

                                setSelectedDatascope(e.target.value)
                                appDispatch( setSelectedDatascopeDefault(e.target.value) )

                            }}
                            value={selectedDatascope}
                        >
                            return <option value="">...</option>
                            {datascopes != null && datascopes.map(x => {

                                return <option key={x.id} value={x.id}>{x.name}</option>
                            })
                            }
                        </Form.Control>

                    </Form.Group>
                </Col>
                </Row>
                */}
                <Row>
                    <Col>
                        <ToolkitProvider
                            bootstrap4
                       
                            keyField='id'
                            data={datascopes}
                            columns={columns}
                            search >
                            {
                                props => (
                                    <div className="text-left">
                                   
                                        <div className="d-flex">
                                            <h5 >Edit Ghost Databases  <i onClick={e => { setShowOffhelp(!showOffhelp) }} className="trash fa-solid fa-circle-info mr-1"></i><Spinner show={spinner} style={{ width: '28px' }}></Spinner></h5>


                                            <div style={{ marginLeft: "auto" }}>
                                                <SearchBar size="sm" {...props.searchProps} />
                                                <ClearSearchButton {...props.searchProps} />
                                                <i onClick={e=>reload()} className="fa fa-refresh ablue cursor-pointer" style={{position: 'relative', top: '2px'}} aria-hidden="true"></i>

                                            </div>
                                        </div>
                                        <div className="d-block">
                                            <BootstrapTable id="datascopetable"
                                                condensed
                                                keyField='id'
                                                selectRow={selectRow}
                                                striped bootstrap4 bordered={false}
                                                pagination={paginationFactory({
                                                    sizePerPage: 4,
                                                    sizePerPageList: [4, 8, 12, 16]
                                                })}
                                                {...props.baseProps}
                                            />
                                        </div>
                                    </div>
                                )
                            }
                        </ToolkitProvider>
                    </Col>
                </Row>
                {(selectedDatascope !== "" && selectedDatascope !== "xxx") &&
                    <div className=" text-left p-4 mt-4" style={{ backgroundColor: "rgba(255, 255, 255, 0.7)" }}>
                        <h5>{selectedDatascopeDetails?.name}</h5>
                        <Form onSubmit={handleSubmit} ref={form} noValidate validated={validated}>
                            <DatascopeForm edit={true} dbname={dbname} onDbname={setDbname}
                                onDeleteConnection={onDeleteConnection}
                                onTablesMapUpdate={onTablesMapUpdate} onEditTable={onEditTable}
                                AddNewTable={addNewTable} onAddTableRef={onAddTableRef} connections={conns}
                                setAlert={setAlert} nameToConnection={remap}
                                initialTables={initialTables}
                            />

                            <Button variant="dymium" size="sm" className="mt-4" type="submit">
                                Apply
                            </Button>
                        </Form>
                    </div>
                }

            </div>
        </div>
    )
}