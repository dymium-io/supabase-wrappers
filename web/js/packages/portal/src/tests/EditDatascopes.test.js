/**
 * @jest-environment jsdom
 */
import React from 'react';
import renderer from 'react-test-renderer';

import { BrowserRouter } from "react-router-dom";
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux'
import store from '../Store/index'
import { Http } from '@dymium/common/Api/Http'
import  EditDatascopes  from '../App/EditDatascopes'
import 'bootstrap/dist/css/bootstrap.min.css';
import '@dymium/common/App.scss';
import "@fortawesome/fontawesome-free/css/all.min.css";
import mockFetch from "../mocks/mockFetch";


beforeEach(() => {
  global.fetch = jest.fn().mockImplementation(() => mockFetch)();
  global.sessionStorage.setItem("Session", "mockJWT");
})

afterEach(() => {
   jest.restoreAllMocks()
});

test('add-datascope', async () => {
    const component = await act( async () => renderer.create(
        <Provider store={store}>
            <BrowserRouter>
                <EditDatascopes />
            </BrowserRouter>
        </Provider >
    ));
    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
});