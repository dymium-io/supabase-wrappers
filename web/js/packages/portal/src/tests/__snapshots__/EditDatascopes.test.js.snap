// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add-datascope 1`] = `
<div
  className=" text-left"
>
  <div
    className="bs-canvas-overlay position-fixed  w-100 h-100"
    id="bs-offcanvas-overlay"
  />
  <div
    className="bs-offcanvas position-fixed  h-100 bs-offcanvas-anim "
    style={
      {
        "right": -600,
        "width": 0,
      }
    }
  >
    <div>
      <div
        className="row"
      >
        <div
          className="col"
        >
           
          <h5>
            Link table
             
          </h5>
        </div>
        <div
          className="col-auto"
        >
          <i
            aria-hidden="true"
            className="cursor-pointer fa fa-times mr-1 fa-2x close"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="offcanvas_handle p-0 m-0 px-2 col-auto"
        onMouseDown={[Function]}
      />
      <div
        className="pl-0 col"
        style={
          {
            "paddingRight": "10px",
          }
        }
      />
    </div>
  </div>
  <div
    className="bs-offcanvas position-fixed  h-100 bs-offcanvas-anim  cthrough"
    style={
      {
        "right": -300,
        "width": 0,
      }
    }
  >
    <div>
      <div
        className="row"
      >
        <div
          className="col"
        >
           
          <h5>
             
          </h5>
        </div>
        <div
          className="col-auto"
        >
          <i
            aria-hidden="true"
            className="cursor-pointer fa fa-times mr-1 fa-2x close"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="offcanvas_handle p-0 m-0 px-2 col-auto"
        onMouseDown={[Function]}
      />
      <div
        className="pl-0 col"
        style={
          {
            "paddingRight": "10px",
          }
        }
      >
        <h5>
          Editing Ghost Database
        </h5>
        <div
          className="mb-3"
        >
          This page allows to edit or remove a Ghost Database.
        </div>
        <div
          className="mb-3"
        >
          The same interface as for adding a Ghost Database is used. You select Data Sources and tables that you want to expose
        </div>
      </div>
    </div>
  </div>
  <div
    className=" text-left"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="text-left"
        >
          <div
            className="d-flex"
          >
            <h5>
              Edit Ghost Databases  
              <i
                className="trash fa-solid fa-circle-info mr-1"
                onClick={[Function]}
              />
              <img
                className="spinner"
                src="/logo.png"
                style={
                  {
                    "width": "28px",
                  }
                }
              />
            </h5>
            <div
              style={
                {
                  "marginLeft": "auto",
                }
              }
            >
              <label
                className="search-label"
                htmlFor="search-bar-0"
              >
                <span
                  className="sr-only"
                  id="search-bar-0-label"
                >
                  Search this table
                </span>
                <input
                  aria-labelledby="search-bar-0-label"
                  className="form-control "
                  id="search-bar-0"
                  onChange={[Function]}
                  onKeyUp={[Function]}
                  placeholder="Search"
                  style={{}}
                  type="text"
                  value=""
                />
              </label>
              <button
                className="btn btn-default "
                onClick={[Function]}
              >
                Clear
              </button>
              <i
                aria-hidden="true"
                className="fa fa-refresh ablue cursor-pointer"
                onClick={[Function]}
                style={
                  {
                    "position": "relative",
                    "top": "2px",
                  }
                }
              />
            </div>
          </div>
          <div
            className="d-block"
          >
            <div
              className="react-bootstrap-table"
            >
              <table
                className="table table-striped table-sm"
                id="datascopetable"
              >
                <thead>
                  <tr>
                    <th
                      className="selection-cell-header"
                      data-row-selection={true}
                    />
                    <th
                      aria-label="Name: sortable"
                      className="sortable"
                      onClick={[Function]}
                      onKeyUp={[Function]}
                      tabIndex={0}
                    >
                      Name:
                      <span
                        className="order-4"
                      />
                    </th>
                    <th
                      aria-label="Created: sortable"
                      className="sortable"
                      onClick={[Function]}
                      onKeyUp={[Function]}
                      style={
                        {
                          "width": "auto",
                        }
                      }
                      tabIndex={0}
                    >
                      Created:
                      <span
                        className="order-4"
                      />
                    </th>
                    <th
                      aria-label="Modified: sort desc"
                      className="sortable"
                      onClick={[Function]}
                      onKeyUp={[Function]}
                      style={
                        {
                          "width": "auto",
                        }
                      }
                      tabIndex={0}
                    >
                      Modified:
                      <span
                        className="caret-4-desc"
                      />
                    </th>
                    <th
                      style={
                        {
                          "width": "50px",
                        }
                      }
                      tabIndex={0}
                    >
                      Edit
                    </th>
                    <th
                      style={
                        {
                          "width": "90px",
                        }
                      }
                      tabIndex={0}
                    >
                      Delete
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr
                    style={{}}
                  >
                    <td
                      className="selection-cell"
                      onClick={[Function]}
                    >
                      <input
                        checked={false}
                        className="selection-input-4"
                        disabled={false}
                        onChange={[Function]}
                        type="radio"
                      />
                    </td>
                    <td
                      className="overflow-x-scroll"
                    >
                      test1
                    </td>
                    <td>
                      <div>
                        Invalid Date
                      </div>
                    </td>
                    <td>
                      <div>
                        Invalid Date
                      </div>
                    </td>
                    <td
                      style={
                        {
                          "height": "30px",
                          "textAlign": "center",
                        }
                      }
                    >
                      <i
                        className="fas fa-edit ablue"
                        onClick={[Function]}
                        role="button"
                      />
                    </td>
                    <td
                      style={
                        {
                          "height": "30px",
                          "textAlign": "center",
                        }
                      }
                    >
                      <i
                        className="fas fa-trash ablue"
                        onClick={[Function]}
                        role="button"
                      />
                    </td>
                  </tr>
                  <tr
                    style={{}}
                  >
                    <td
                      className="selection-cell"
                      onClick={[Function]}
                    >
                      <input
                        checked={false}
                        className="selection-input-4"
                        disabled={false}
                        onChange={[Function]}
                        type="radio"
                      />
                    </td>
                    <td
                      className="overflow-x-scroll"
                    >
                      test2
                    </td>
                    <td>
                      <div>
                        Invalid Date
                      </div>
                    </td>
                    <td>
                      <div>
                        Invalid Date
                      </div>
                    </td>
                    <td
                      style={
                        {
                          "height": "30px",
                          "textAlign": "center",
                        }
                      }
                    >
                      <i
                        className="fas fa-edit ablue"
                        onClick={[Function]}
                        role="button"
                      />
                    </td>
                    <td
                      style={
                        {
                          "height": "30px",
                          "textAlign": "center",
                        }
                      }
                    >
                      <i
                        className="fas fa-trash ablue"
                        onClick={[Function]}
                        role="button"
                      />
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div
              className="row react-bootstrap-table-pagination"
            >
              <div
                className="col-md-6 col-xs-6 col-sm-6 col-lg-6"
              >
                <span
                  className="react-bs-table-sizePerPage-dropdown dropdown"
                  style={
                    {
                      "visibility": "visible",
                    }
                  }
                >
                  <button
                    aria-expanded={false}
                    className="btn btn-default btn-secondary dropdown-toggle"
                    data-toggle="dropdown"
                    id="datascopetable-pageDropDown"
                    onBlur={[Function]}
                    onClick={[Function]}
                    type="button"
                  >
                    5
                     
                  </button>
                  <ul
                    aria-labelledby="datascopetable-pageDropDown"
                    className="dropdown-menu "
                    role="menu"
                  >
                    <a
                      className="dropdown-item"
                      data-page={5}
                      href="#"
                      onMouseDown={[Function]}
                      role="menuitem"
                      tabIndex="-1"
                    >
                      5
                    </a>
                    <a
                      className="dropdown-item"
                      data-page={10}
                      href="#"
                      onMouseDown={[Function]}
                      role="menuitem"
                      tabIndex="-1"
                    >
                      10
                    </a>
                    <a
                      className="dropdown-item"
                      data-page={15}
                      href="#"
                      onMouseDown={[Function]}
                      role="menuitem"
                      tabIndex="-1"
                    >
                      15
                    </a>
                    <a
                      className="dropdown-item"
                      data-page={20}
                      href="#"
                      onMouseDown={[Function]}
                      role="menuitem"
                      tabIndex="-1"
                    >
                      20
                    </a>
                  </ul>
                </span>
              </div>
              <div
                className="react-bootstrap-table-pagination-list col-md-6 col-xs-6 col-sm-6 col-lg-6"
              >
                <ul
                  className="pagination react-bootstrap-table-page-btns-ul"
                >
                  <li
                    className="active page-item"
                    title="1"
                  >
                    <a
                      className="page-link"
                      href="#"
                      onClick={[Function]}
                    >
                      1
                    </a>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
